### DISCUSSION TITLE
Pellet Level Sensor

### DISCUSSION THREAD

**Discussion by nebhead** [nebhead/PiFire#17] [Ideas]
Hi all - I'm considering adding a pellet level sensor to monitor the pellet level in the hopper to notify the user when the pellets are starting to get low.  I just ordered a VL53L0X Time of Flight sensor which I'm going to test out soon.  Thought it might be cool to see if anyone had any input on this to see if there are better sensor options out there.

**Comment by faisalhusain007** [comment]
Sounds good, could use the sensor and the distance it detects to show a graphical representation of how full the hopper is. I think it is better than using a switch or load cell which will be harder to fit and require calibration before use. I think it's a good solution. The HC-SR04 ultrasonic sensor has a wide field of view so not sure how accurate it would be, but really cheap.

**Comment by nebhead** [comment]
Cool - I have a HC-SR04, but I was a little worried about both the size and the fact that it's going to be in a rather dusty environment (inside the hopper).  I suppose the VL53L0X might be susceptible to dust as well, but maybe less so?  Also, it can be hooked up via the I2C which is kinda nice.

**Comment by weberbox** [comment]
So I was designing a mount for the pellet sensor but was wondering how anyone else is positioning theirs. 

Would it be best be right over the auger opening where the pellets make a cone as they go down or somewhere else? 

Are you attaching it to the lid or somewhere else? 


My grill has a metal grid as a safety for the pellet box. It is annoying when trying to swap pellets or even just moving pellets around. I originally wanted to mount the sensor on this grate but I am rethinking that as I temporarily removed the grate and it is so much nicer to deal with the pellets now. Also I usually fill the pellets above this grate which is not a huge deal but the sensor would not work until they go below it.  

Next would be some sort of mounting to the lid but there is no great area for that and the wires would need to bent or move every open which could lead to issues. Also it would require drilling holes for screws or using some sort of double sided tape that might not stick over time due to heat 


Any pictures would be awesome

**Reply by nebhead** [reply]
I'm also looking for any good ideas for this.  I was wanting to at the very least 3D print an enclosure for the sensor, but I just haven't gotten around to doing that yet.  

My current solution is some double-sided tape, which I have taped to the lid of the hopper (right over the auger hole area).  Also, I've only used the tape on the Dupont connectors, so that the device sort of just hangs down.  So far this, admittedly janky, solution is actually working well for me.  It stays out of the way when I'm loading pellets, and hasn't caused any problems thus far.  It just doesn't look pretty.  Here's what it looks like: 

![image](https://user-images.githubusercontent.com/19876886/129410211-a22521b2-5ebf-4a17-8277-619b2e3ae9e1.png)

One other thing, _really_ need to have two measurements for the hopper calculations.  One for 'empty' (which is already there) and one for 'full'.  As it stands, my full hopper still measures about 4cm distance and that means that I will never have a technically full hopper.  So, my plan is to move the empty/full settings to settings and allow those to be customized from the settings screen.  Just haven't gotten around to it yet.

**Reply by nebhead** [reply]
FYI... pushed an update to the development branch to improve hopper calculations by adding "full" value.  Also added these hopper settings to the settings page so that they can be configured easily.  

https://github.com/nebhead/PiFire/commit/6e7ae3605ad1b874b867376d7cc80b402dc7bc61

As mentioned in the commit, will probably break your current install, so you'll need to either edit or delete pelletdb.json before restarting.  There may be other breaking changes - so your mileage may vary.  Hence, proceed with caution. :)

**Reply by weberbox** [reply]
Finally got my pellet level sensor installed. Still need to calibrate the full and empty measurements but it is a great addition. 

[Cad Design](https://cad.onshape.com/documents/b95282d5ebd233bf22c48187/w/867e1e2e2f53551dd2575f1a/e/2d386f3c04c0d6736fcd800f)

![20211017_162454.jpg](https://user-images.githubusercontent.com/313328/137690474-e3ac5deb-2487-4275-9768-5563362a0425.jpg)
![20211017_162501.jpg](https://user-images.githubusercontent.com/313328/137690500-979ebcd1-f1d6-4692-8817-0e649a8b93b8.jpg)
![20211017_162516.jpg](https://user-images.githubusercontent.com/313328/137690517-6f6aa09b-ebe3-41a5-89db-d479b2777cdc.jpg)
